staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -340,3 +340,9 @@ func (pf *PortForwarder) Close() {
 		}
 	}
 }
+
+// GetPorts will return the ports that were forwarded; this can be used to
+// retrieve the locally-bound port in cases where the input was port 0.
+func (pf *PortForwarder) GetPorts() []ForwardedPort {
+	return pf.ports
```

Comment:
calls to this for purposes of retrieving auto-assigned port numbers are necessarily made from a separate goroutine than the one that called `ForwardPorts()` (since that method blocks until the operation is complete). That means that anything using the value returned from this method is subject to data races with the `port.Local = uint16(localPortUInt)` assignment.

At the very least, we need to protect against that data race.

A bigger question is whether this method should block until the `ForwardPorts()` call has established listeners and the actual local ports are available.

Reply:
or, could require a `Ready` channel be specified in order to use GetPorts() (which they have to provide if they want to call GetPorts() safely)... what do you think of something like this:

```go
func (pf *PortForwarder) GetPorts() ([]ForwardedPort, error) {
	if pf.Ready == nil {
		return nil, fmt.Errorf("no Ready channel provided")
	}
	select {
	case <-pf.Ready:
		return pf.ports, nil
	default:
		return nil, fmt.Errorf("listeners not ready")
	}
}
```

Reply:
SGTM. Updated the PR.